src/AspNetCore/Src/inprocessapplication.cxx

```diff
@@ -0,0 +1,780 @@
+)
+{
+    return pHttpContext->GetRequest()->GetRawHttpRequest();
+}
+
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+HTTP_RESPONSE*
+http_get_raw_response(
+    _In_ IHttpContext* pHttpContext
+)
+{
+    return pHttpContext->GetResponse()->GetRawHttpResponse();
+}
+
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT VOID http_set_response_status_code(
+    _In_ IHttpContext* pHttpContext,
+    _In_ USHORT statusCode,
+    _In_ PCSTR pszReason
+)
+{
+    pHttpContext->GetResponse()->SetStatus(statusCode, pszReason);
+}
+
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+HRESULT
+http_post_completion(
+    _In_ IHttpContext* pHttpContext
+)
+{
+    return pHttpContext->PostCompletion(0);
+}
+
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+VOID
+http_indicate_completion(
+    _In_ IHttpContext* pHttpContext,
+    _In_ REQUEST_NOTIFICATION_STATUS notificationStatus
+)
+{
+    pHttpContext->IndicateCompletion(notificationStatus);
+}
+
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+VOID
+http_get_completion_info(
+    _In_ IHttpCompletionInfo2* info,
+    _Out_ DWORD* cbBytes,
+    _Out_ HRESULT* hr
+)
+{
+    *cbBytes = info->GetCompletionBytes();
+    *hr = info->GetCompletionStatus();
+}
+
+//
+// todo: we should not rely on INPROCESS_APPLICATION::GetInstance()
+// the signature should be changed. application's based address should be passed in
+//
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+BSTR // TODO probably should make this a wide string
+http_get_application_full_path()
+{
+    LPWSTR pwzPath = NULL;
+    INPROCESS_APPLICATION* pApplication = INPROCESS_APPLICATION::GetInstance();
+    if(pApplication != NULL)
+    {
+        pwzPath = pApplication->QueryConfig()->QueryApplicationFullPath()->QueryStr();
+    }
+    return SysAllocString(pwzPath);
+}
+
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+HRESULT
+http_read_request_bytes(
+    _In_ IHttpContext* pHttpContext,
+    _In_ CHAR* pvBuffer,
+    _In_ DWORD cbBuffer,
+    _In_ PFN_ASYNC_COMPLETION pfnCompletionCallback,
+    _In_ VOID* pvCompletionContext,
+    _In_ DWORD* pDwBytesReceived,
+    _In_ BOOL* pfCompletionPending
+)
+{
+    IHttpRequest3 *pHttpRequest = (IHttpRequest3*)pHttpContext->GetRequest();
+
+    BOOL fAsync = TRUE;
+
+    HRESULT hr = pHttpRequest->ReadEntityBody(
+        pvBuffer,
+        cbBuffer,
+        fAsync,
+        pfnCompletionCallback,
+        pvCompletionContext,
+        pDwBytesReceived,
+        pfCompletionPending);
+
+    if (hr == HRESULT_FROM_WIN32(ERROR_HANDLE_EOF))
+    {
+        // We reached the end of the data
+        hr = S_OK;
+    }
+
+    return hr;
+}
+
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+HRESULT
+http_write_response_bytes(
+    _In_ IHttpContext* pHttpContext,
+    _In_ HTTP_DATA_CHUNK* pDataChunks,
+    _In_ DWORD nChunks,
+    _In_ PFN_ASYNC_COMPLETION pfnCompletionCallback,
+    _In_ VOID* pvCompletionContext,
+    _In_ BOOL* pfCompletionExpected
+)
+{
+    IHttpResponse2 *pHttpResponse = (IHttpResponse2*)pHttpContext->GetResponse();
+
+    BOOL fAsync = TRUE;
+    BOOL fMoreData = TRUE;
```

Comment:
Should we be sending fMoreData from managed to native?

Reply:
It seems to be a pattern like IIS CGI handler does. This layer does not know whether there will be more data send. Only App knows about it. It is OK to assume to have more data as cost of one more call.

Reply:
Yes it should be passed